Add alias for creating GitHub Releases

ap0llo opened the pull request:
## Description

This PR adds the alias `GitHubCreateReleaseAsync()` which allows creating a GitHub Release from cake,
as discussed in https://github.com/ap0llo/Cake.GitHubReleases/issues/1.

Since the code I brought over from [ap0llo/Cake.GitHubReleases](https://github.com/ap0llo/Cake.GitHubReleases) was already using nullable reference types, I enabled nullable reference types  (and C# 8) for the entire project and added nullable annotations to the existing code.

## Related Issue

- https://github.com/ap0llo/Cake.GitHubReleases/issues/1, 

## Motivation and Context

With this change, it will be possible to create a GitHub Release from a Cake script.
The alias simplifies creation of the release without the need to use Octokit directly.

Integrating this functionality into Cake.GitHub rather than maintaining a separate addin should make it easier for users to integrate GitHub functionality into their Cake script and also avoids potential dependency conflicts that might arise from two addins depending on different versions of Octokit.

## How Has This Been Tested?

Almost all of the functionality for creating releases is implemented in the `GitHubReleaseCreator` class.
I added unit tests to the existing `Cake.GitHub.Tests` project that cover all scenarios (at least all of the scenarios I could think of). The tests use Moq to mock the access to GitHub.


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project:
*I tried to make the signature of `GitHubCreateReleaseAsync()` as similar to the existing `GitHubStatus()` alias to make both of them as consistent as possible. If you have additional notes w.r.t. code style, please let me know*
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

GeertvanHorrik commented:
Not sure why I missed this one. Will review asap.

GeertvanHorrik commented on src/Cake.GitHub.Tests/Helpers/MoqExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+    internal static class MoqExtensions
+    {
+        public static IReturnsResult<TMock> ThrowsNotFoundAsync<TMock, TResult>(this IReturns<TMock, Task<TResult>> mock) where TMock : class =>
+            mock.ThrowsAsync(new NotFoundException("", HttpStatusCode.NotFound));
```
This is not wrong, but I prefer string.Empty

ap0llo replied:
Sure, I've updated it to use string.Empty

GeertvanHorrik commented on src/Cake.GitHub.Tests/Helpers/GitHubClientMock.cs:
```diff
@@ -0,0 +1,58 @@
+    {
+        public class RepositoriesClientMock
+        {
+            private readonly Mock<IRepositoriesClient> m_Mock = new Mock<IRepositoriesClient>(MockBehavior.Strict);
```
I think we don't use the `m_` prefix anywhere else?

ap0llo replied:
That convention is used in the code bases I usually work on, so I just used it without thinking about it, I've updated all fields to use just a `_` prefix

GeertvanHorrik commented on src/Cake.GitHub.Tests/Helpers/XunitCakeLog.cs:
```diff
@@ -0,0 +1,31 @@
+    /// </summary>
+    internal class XunitCakeLog : ICakeLog
+    {
+        private readonly ITestOutputHelper m_TestOutputHelper;
```
`_testOutputHelper` instead?

ap0llo replied:
Fixed

GeertvanHorrik commented on src/Cake.GitHub/GitHubAppAliases.Releases.cs:
```diff
@@ -0,0 +1,49 @@
+        [CakeMethodAlias]
+        public static async Task<GitHubRelease> GitHubCreateReleaseAsync(
+            this ICakeContext context,
+            string userName,
```
should this be nullable, similar to what is mentioned in the docs?

ap0llo replied:
You're right

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubRelease.cs:
```diff
@@ -0,0 +1,126 @@
+    /// </summary>
+    public sealed class GitHubRelease
+    {
+        private readonly List<GitHubReleaseAsset> m_Assets = new List<GitHubReleaseAsset>();
```
use `_assets` instead?

ap0llo replied:
Fixed

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubRelease.cs:
```diff
@@ -0,0 +1,126 @@
+
+        internal void Add(GitHubReleaseAsset asset)
+        {
+            if (asset is null)
```
Can this be null when we have nullable enabled on the project?

ap0llo replied:
You're right, we can omit the null checks for internal methods, I've updated the code accordingly

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseAsset.cs:
```diff
@@ -0,0 +1,72 @@
+        /// </summary>
+        internal static GitHubReleaseAsset FromReleaseAsset(ReleaseAsset asset)
+        {
+            if (asset is null)
```
no need to check?

ap0llo replied:
Fixed

GeertvanHorrik reviewed (commented):
First of all, sorry for my late review, the notification got buried in my inbox. This code looks good, I've reviewed most of it and added some comments. Looking forward to hearing your opinion about it.

Once I know what direction we are going with the current comments, we can finalize the review and merge this.

ap0llo commented:
> First of all, sorry for my late review, the notification got buried in my inbox. 

 No worries :)

> This code looks good, I've reviewed most of it and added some comments. Looking forward to hearing your opinion about it.
> Once I know what direction we are going with the current comments, we can finalize the review and merge this.

All your comments are reasonable, I've updated the code accordingly.

Would you like me to squash the commits together a bit or should I just leave the cleanup commits in place?

GeertvanHorrik commented:
> Would you like me to squash the commits together a bit or should I just leave the cleanup commits in place?

My preference is to keep it, even for PRs the history of commits can be very useful to understand the "thought process" afterwards. So if you agree, let's keep it.

Will do a re-review as soon as I can.

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubRelease.cs:
```diff
@@ -0,0 +1,120 @@
+        /// </summary>
+        internal static GitHubRelease FromRelease(Release release)
+        {
+            if (release is null)
```
cannot be null

ap0llo replied:
Fixed

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseCreator.cs:
```diff
@@ -0,0 +1,216 @@
+        
+        public GitHubReleaseCreator(ICakeLog cakeLog, IFileSystem fileSystem, IGitHubClient githubClient)
+        {
+            _cakeLog = cakeLog ?? throw new ArgumentNullException(nameof(cakeLog));
```
the throws can be removed as well, nothing will ever be null here.

ap0llo replied:
Fixed

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseCreator.cs:
```diff
@@ -0,0 +1,216 @@
+            if (String.IsNullOrWhiteSpace(tagName))
+                throw new ArgumentException("Value must not be null or whitespace", nameof(tagName));
+
+            if (settings is null)
```
this can be removed

ap0llo replied:
Fixed

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseCreator.cs:
```diff
@@ -0,0 +1,216 @@
+            // Check for existing release and delete if necessary
+            //
+            var existingRelease = await TryGetReleaseAsync(owner: owner, repository: repository, tagName: tagName);
+            if (existingRelease != null)
```
I think "is not null" is preferred nowadays

ap0llo replied:
The `is not null` pattern requires C# 9 which won't work on the Visual Studio 2017 build machines currently used for CI.
I didn't want to change any unrelated config as part of a feature PR

GeertvanHorrik replied:
Good point.

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseCreator.cs:
```diff
@@ -0,0 +1,216 @@
+            var existingRelease = await TryGetReleaseAsync(owner: owner, repository: repository, tagName: tagName);
+            if (existingRelease != null)
+            {
+                if (settings.Overwrite)
```
just for readability: if we invert this, it saves us reading / understanding the code:

```
if (!settings.Overwrite)
{
     throw new Release ....
}

await DeleteReleaseAsync...
```

ap0llo replied:
Fixed

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseCreator.cs:
```diff
@@ -0,0 +1,216 @@
+                var matchingReleases = allReleases.Where(x => StringComparer.Ordinal.Equals(x.TagName, tagName)).ToArray();
+
+                Release? release;
+                if(matchingReleases.Length == 0)
```
can be replaced with SingleOrDefault()?

ap0llo replied:
I intentionally did not use `SingleOrDefault()` to get a better exception message rather than the generic message given by `SingleOrDefault()` and catching the exception just to rethrow it seemed wrong as well

GeertvanHorrik replied:
Makes sense, thanks for explaining.

GeertvanHorrik commented on src/Cake.GitHub/Releases/GitHubReleaseCreator.cs:
```diff
@@ -0,0 +1,216 @@
+                    throw new AmbiguousTagNameException($"There are multiple existing releases for tag name '{tagName}'");
+                }
+                
+                if (release != null)
```
is not null

ap0llo replied:
This would require C# 9 (see above)

GeertvanHorrik reviewed (commented):
Reviewed everything except the unit tests. Let's discuss the current comments and see if we both agree on them.

Then we can hopefully finalize this PR because it looks great so far 👍

GeertvanHorrik commented:
@pascalberger we might need your help here. 

@ap0llo did a great job adding this PR, and I've reviewed (and approved). Do we need to synchronize any other files to make this work (e.g. cake scripts, recipes, etc)?

gep13 commented:
>@GeertvanHorrik said...
Do we need to synchronize any other files to make this work (e.g. cake scripts, recipes, etc)?

What changes are you thinking need to happen here?

GeertvanHorrik commented:
I saw the check failing, and it's been a while we synced the recipe files, so I made the assumption it might be time for a synchronization. Can we just merge this PR?

gep13 commented:
Oh, I see what you mean...

There is a known issue that I have seen with running GitVersion on AppVeyor when it is a Pull Request.  I am not entirely sure what is going on there, as I haven't dug into it.  It could be that this is solved with a newer version of GitVersion, but as mentioned, I haven't dug into it.

If the build is working locally for you, then I would recommend merging the PR, and in all likelihood it will succeed currently when building off the develop branch.

GeertvanHorrik commented:
Thanks for the explanation, going to merge now.
